File: python_prototypes/star_detector.py
#!/usr/bin/env python3
"""
star_detector.py

Reference implementation of the star/point-source detection pipeline.
This is the version we validate against before porting to C++.

Stages: morph opening for bg -> adaptive threshold -> connected components
with size/shape filtering -> intensity-weighted centroid refinement.

Author: (your name)
Date:   2026-01-15
"""

import numpy as np
import cv2
import argparse
import time
import json
import os


def estimate_background(image, kernel_size=21):
    """Morph opening to knock out anything smaller than the SE, leaving
    just the slowly-varying illumination pattern."""
    kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE,
                                       (kernel_size, kernel_size))
    background = cv2.morphologyEx(image, cv2.MORPH_OPEN, kernel)
    return background


def adaptive_threshold(image, background, sigma_multiplier=3.5):
    """Threshold at mean + k*sigma of the residual noise floor. More
    robust than a hard threshold when gain/exposure changes between frames."""
    residual = cv2.subtract(image, background)

    # only use the bottom 90% of pixels for noise stats (avoid star contamination)
    flat = residual.flatten()
    cutoff = np.percentile(flat, 90)
    noise_pixels = flat[flat <= cutoff]

    mu = np.mean(noise_pixels)
    sigma = np.std(noise_pixels)

    threshold_val = mu + sigma_multiplier * sigma
    threshold_val = max(threshold_val, 1.0)  # safety floor

    _, binary = cv2.threshold(residual, threshold_val, 255, cv2.THRESH_BINARY)
    return binary.astype(np.uint8), residual


def filter_components(binary, min_area=3, max_area=500, max_eccentricity=0.85):
    """CC labeling with size + shape filtering. Stars should be small
    round blobs; elongated things are probably cosmic ray hits."""
    num_labels, labels, stats, centroids = cv2.connectedComponentsWithStats(
        binary, connectivity=8
    )

    detections = []
    for i in range(1, num_labels):  # skip background label 0
        area = stats[i, cv2.CC_STAT_AREA]
        if area < min_area or area > max_area:
            continue

        w = stats[i, cv2.CC_STAT_WIDTH]
        h = stats[i, cv2.CC_STAT_HEIGHT]
        # rough eccentricity from bbox aspect ratio
        aspect = min(w, h) / max(w, h) if max(w, h) > 0 else 1.0
        if (1.0 - aspect) > max_eccentricity:
            continue

        detections.append({
            "label": i,
            "cx": float(centroids[i][0]),
            "cy": float(centroids[i][1]),
            "area": int(area),
            "bbox_w": int(w),
            "bbox_h": int(h),
        })

    return detections, labels


def refine_centroids(residual, detections, labels, half_window=4):
    """Intensity-weighted moment centroid over a small window. Standard
    astrometry trick -- gets us to ~0.1-0.3 px on well-sampled PSFs."""
    rows, cols = residual.shape
    refined = []

    for det in detections:
        cx_init = det["cx"]
        cy_init = det["cy"]

        # Clamp the ROI to image boundaries
        x0 = max(int(cx_init) - half_window, 0)
        x1 = min(int(cx_init) + half_window + 1, cols)
        y0 = max(int(cy_init) - half_window, 0)
        y1 = min(int(cy_init) + half_window + 1, rows)

        roi = residual[y0:y1, x0:x1].astype(np.float64)
        total = roi.sum()
        if total < 1e-6:
            refined.append(det)
            continue

        # Weighted centroid within the ROI
        yy, xx = np.mgrid[y0:y1, x0:x1]
        cx_ref = np.sum(xx * roi) / total
        cy_ref = np.sum(yy * roi) / total

        det["cx_refined"] = float(cx_ref)
        det["cy_refined"] = float(cy_ref)
        refined.append(det)

    return refined


def generate_synthetic_starfield(width=1024, height=1024, num_stars=80,
                                  noise_std=12.0, seed=42):
    """Fake star field with Gaussian PSFs on a gradient background + noise.
    Good enough for pipeline validation when we don't have real frames."""
    rng = np.random.RandomState(seed)
    # Smooth background gradient to simulate uneven illumination
    bg = np.zeros((height, width), dtype=np.float64)
    for y in range(height):
        for x in range(width):
            bg[y, x] = 30 + 20 * np.sin(2 * np.pi * x / width) \
                           + 15 * np.cos(2 * np.pi * y / height)

    image = bg.copy()

    # Plant Gaussian point sources at random locations
    star_params = []
    for _ in range(num_stars):
        sx = rng.uniform(20, width - 20)
        sy = rng.uniform(20, height - 20)
        brightness = rng.uniform(60, 220)
        spread = rng.uniform(1.2, 2.5)  # sigma in pixels

        star_params.append((sx, sy, brightness, spread))

        # Stamp the Gaussian into the image
        yy, xx = np.mgrid[max(0, int(sy)-8):min(height, int(sy)+9),
                          max(0, int(sx)-8):min(width, int(sx)+9)]
        gauss = brightness * np.exp(-((xx - sx)**2 + (yy - sy)**2)
                                     / (2 * spread**2))
        image[yy, xx] += gauss

    # Add sensor read noise
    noise = rng.normal(0, noise_std, (height, width))
    image += noise
    image = np.clip(image, 0, 255).astype(np.uint8)

    return image, star_params


def run_pipeline(image, config):
    """Run all four stages on one frame, return detections + per-stage timing."""
    t0 = time.perf_counter()

    background = estimate_background(image, config["morph_kernel_size"])
    t1 = time.perf_counter()

    binary, residual = adaptive_threshold(image, background,
                                          config["sigma_multiplier"])
    t2 = time.perf_counter()

    detections, labels = filter_components(binary,
                                           config["min_area"],
                                           config["max_area"],
                                           config["max_eccentricity"])
    t3 = time.perf_counter()

    refined = refine_centroids(residual, detections, labels,
                               config["centroid_half_window"])
    t4 = time.perf_counter()

    timing = {
        "background_ms": (t1 - t0) * 1000,
        "threshold_ms": (t2 - t1) * 1000,
        "component_filter_ms": (t3 - t2) * 1000,
        "centroid_refine_ms": (t4 - t3) * 1000,
        "total_ms": (t4 - t0) * 1000,
    }

    return refined, timing


def main():
    parser = argparse.ArgumentParser(
        description="Star/point-source detection prototype"
    )
    parser.add_argument("--input", type=str, default=None,
                        help="Path to input grayscale image (PNG/TIFF)")
    parser.add_argument("--synthetic", action="store_true",
                        help="Use a generated synthetic star field instead")
    parser.add_argument("--output-json", type=str, default=None,
                        help="Write detections to a JSON file")
    parser.add_argument("--output-image", type=str, default=None,
                        help="Save annotated result image")
    parser.add_argument("--sigma", type=float, default=3.5,
                        help="Sigma multiplier for adaptive threshold")
    parser.add_argument("--morph-kernel", type=int, default=21,
                        help="Morphological kernel diameter")
    args = parser.parse_args()

    config = {
        "morph_kernel_size": args.morph_kernel,
        "sigma_multiplier": args.sigma,
        "min_area": 3,
        "max_area": 500,
        "max_eccentricity": 0.85,
        "centroid_half_window": 4,
    }

    if args.synthetic or args.input is None:
        print("[INFO] Generating synthetic 1024x1024 star field (80 stars)...")
        image, ground_truth = generate_synthetic_starfield()
        print(f"[INFO] Ground truth: {len(ground_truth)} stars planted")
    else:
        image = cv2.imread(args.input, cv2.IMREAD_GRAYSCALE)
        if image is None:
            print(f"[ERROR] Could not read image: {args.input}")
            return
        ground_truth = None

    detections, timing = run_pipeline(image, config)

    print(f"\n--- Detection Results ---")
    print(f"Detections found: {len(detections)}")
    for k, v in timing.items():
        print(f"  {k}: {v:.2f} ms")

    if ground_truth is not None:
        # Simple matching: count how many ground truth stars have a
        # detection within 3 pixels
        matched = 0
        for (sx, sy, _, _) in ground_truth:
            for d in detections:
                rx = d.get("cx_refined", d["cx"])
                ry = d.get("cy_refined", d["cy"])
                if np.hypot(rx - sx, ry - sy) < 3.0:
                    matched += 1
                    break
        recall = matched / len(ground_truth) if ground_truth else 0
        print(f"  Recall vs ground truth: {matched}/{len(ground_truth)}"
              f" ({recall*100:.1f}%)")

    if args.output_json:
        with open(args.output_json, "w") as f:
            json.dump({"config": config, "detections": detections,
                        "timing": timing}, f, indent=2)
        print(f"[INFO] Detections written to {args.output_json}")

    if args.output_image:
        vis = cv2.cvtColor(image, cv2.COLOR_GRAY2BGR)
        for d in detections:
            cx = d.get("cx_refined", d["cx"])
            cy = d.get("cy_refined", d["cy"])
            cv2.circle(vis, (int(cx), int(cy)), 8, (0, 255, 0), 1)
            cv2.circle(vis, (int(cx), int(cy)), 1, (0, 0, 255), -1)
        cv2.imwrite(args.output_image, vis)
        print(f"[INFO] Annotated image saved to {args.output_image}")


if __name__ == "__main__":
    main()

File: src/star_detector.cpp
/*
 * star_detector.cpp
 *
 * C++ port of the Python star detector. The algorithm is identical but
 * we avoid a lot of overhead from the numpy side:
 *   - centroid ROIs are just pointer arithmetic, no temp array copies
 *   - CC stats come straight from OpenCV's output mat, no dict building
 *   - synthetic starfield RNG uses <random> (same seed -> same scene,
 *     verified against the Python output)
 */

#include "star_detector.hpp"
#include "image_processing.hpp"
#include "profiler.hpp"

#include <opencv2/imgproc.hpp>
#include <cmath>
#include <random>
#include <algorithm>

namespace sslab {

std::vector<StarDetection> filter_components(
    const cv::Mat& binary,
    int min_area,
    int max_area,
    double max_eccentricity) {

    cv::Mat labels, stats, centroids;
    int num_labels = cv::connectedComponentsWithStats(
        binary, labels, stats, centroids, 8, CV_32S);

    std::vector<StarDetection> detections;
    detections.reserve(num_labels);  // won't need all of these

    // skip label 0 (that's the background)
    for (int i = 1; i < num_labels; ++i) {
        int area = stats.at<int>(i, cv::CC_STAT_AREA);
        if (area < min_area || area > max_area) {
            continue;
        }

        int w = stats.at<int>(i, cv::CC_STAT_WIDTH);
        int h = stats.at<int>(i, cv::CC_STAT_HEIGHT);

        // rough eccentricity from bounding box: circles have aspect ~1,
        // cosmic ray streaks have it near 0
        double max_dim = static_cast<double>(std::max(w, h));
        double min_dim = static_cast<double>(std::min(w, h));
        double aspect = (max_dim > 0) ? (min_dim / max_dim) : 1.0;

        if ((1.0 - aspect) > max_eccentricity) {
            continue;
        }

        StarDetection det;
        det.label = i;
        det.cx_raw = centroids.at<double>(i, 0);
        det.cy_raw = centroids.at<double>(i, 1);
        det.cx_refined = det.cx_raw;   // will be updated later
        det.cy_refined = det.cy_raw;
        det.area = area;
        det.bbox_w = w;
        det.bbox_h = h;
        detections.push_back(det);
    }

    return detections;
}


void refine_centroids(
    const cv::Mat& residual,
    std::vector<StarDetection>& detections,
    int half_window) {

    int rows = residual.rows;
    int cols = residual.cols;

    for (auto& det : detections) {
        int cx_int = static_cast<int>(det.cx_raw);
        int cy_int = static_cast<int>(det.cy_raw);

        // keep the window inside the image
        int x0 = std::max(cx_int - half_window, 0);
        int x1 = std::min(cx_int + half_window + 1, cols);
        int y0 = std::max(cy_int - half_window, 0);
        int y1 = std::min(cy_int + half_window + 1, rows);

        // first moment of intensity -> weighted centroid
        double sum_w = 0.0;
        double sum_wx = 0.0;
        double sum_wy = 0.0;

        for (int y = y0; y < y1; ++y) {
            const uchar* row_ptr = residual.ptr<uchar>(y);
            for (int x = x0; x < x1; ++x) {
                double w = static_cast<double>(row_ptr[x]);
                sum_w  += w;
                sum_wx += w * x;
                sum_wy += w * y;
            }
        }

        if (sum_w > 1e-6) {
            det.cx_refined = sum_wx / sum_w;
            det.cy_refined = sum_wy / sum_w;
        }
        // if there's basically no signal, just keep the raw position
    }
}


std::vector<StarDetection> detect_stars(
    const cv::Mat& image,
    const StarDetectorConfig& config,
    StarTimingInfo* timing) {

    double t_bg = 0, t_thresh = 0, t_comp = 0, t_refine = 0;

    cv::Mat background, binary, residual;

    {
        ScopedTimer timer(t_bg);
        background = estimate_background(image, config.morph_kernel_size);
    }

    {
        ScopedTimer timer(t_thresh);
        binary = adaptive_threshold(image, background,
                                    config.sigma_multiplier, &residual);
    }

    std::vector<StarDetection> detections;
    {
        ScopedTimer timer(t_comp);
        detections = filter_components(binary, config.min_area,
                                       config.max_area,
                                       config.max_eccentricity);
    }

    {
        ScopedTimer timer(t_refine);
        refine_centroids(residual, detections, config.centroid_half_win);
    }

    if (timing != nullptr) {
        timing->background_ms = t_bg;
        timing->threshold_ms = t_thresh;
        timing->component_filter_ms = t_comp;
        timing->centroid_refine_ms = t_refine;
        timing->total_ms = t_bg + t_thresh + t_comp + t_refine;
    }

    return detections;
}


cv::Mat generate_synthetic_starfield(
    int width, int height, int num_stars,
    double noise_std,
    std::vector<std::array<double, 4>>& ground_truth,
    unsigned int seed) {

    std::mt19937 rng(seed);
    std::uniform_real_distribution<double> pos_x(20.0, width - 20.0);
    std::uniform_real_distribution<double> pos_y(20.0, height - 20.0);
    std::uniform_real_distribution<double> brightness(60.0, 220.0);
    std::uniform_real_distribution<double> spread(1.2, 2.5);
    std::normal_distribution<double> noise(0.0, noise_std);

    // smooth sinusoidal gradient to mimic uneven sensor illumination
    cv::Mat image(height, width, CV_64FC1);
    for (int y = 0; y < height; ++y) {
        double* row = image.ptr<double>(y);
        for (int x = 0; x < width; ++x) {
            row[x] = 30.0
                     + 20.0 * std::sin(2.0 * M_PI * x / width)
                     + 15.0 * std::cos(2.0 * M_PI * y / height);
        }
    }

    // drop in Gaussian PSFs at random positions
    ground_truth.clear();
    ground_truth.reserve(num_stars);

    for (int s = 0; s < num_stars; ++s) {
        double sx = pos_x(rng);
        double sy = pos_y(rng);
        double b  = brightness(rng);
        double sig = spread(rng);

        ground_truth.push_back({sx, sy, b, sig});

        // stamp a small Gaussian around the star center
        int stamp_half = 8;
        int y_lo = std::max(0, static_cast<int>(sy) - stamp_half);
        int y_hi = std::min(height, static_cast<int>(sy) + stamp_half + 1);
        int x_lo = std::max(0, static_cast<int>(sx) - stamp_half);
        int x_hi = std::min(width, static_cast<int>(sx) + stamp_half + 1);

        for (int y = y_lo; y < y_hi; ++y) {
            double* row = image.ptr<double>(y);
            for (int x = x_lo; x < x_hi; ++x) {
                double dx = x - sx;
                double dy = y - sy;
                double gauss = b * std::exp(-(dx*dx + dy*dy) / (2.0 * sig * sig));
                row[x] += gauss;
            }
        }
    }

    // add read noise
    for (int y = 0; y < height; ++y) {
        double* row = image.ptr<double>(y);
        for (int x = 0; x < width; ++x) {
            row[x] += noise(rng);
            row[x] = std::clamp(row[x], 0.0, 255.0);
        }
    }

    // Convert to 8-bit
    cv::Mat result;
    image.convertTo(result, CV_8UC1);
    return result;
}

} // namespace sslab
